src/goto-analyzer/show_on_source.cpp

```diff
@@ -0,0 +1,156 @@
+/*******************************************************************\
+
+Module: goto-analyzer
+
+Author: Daniel Kroening, kroening@kroening.com
+
+\*******************************************************************/
+
+#include "show_on_source.h"
+
+#include <util/file_util.h>
+#include <util/message.h>
+#include <util/unicode.h>
+
+#include <analyses/ai.h>
+
+#include <fstream>
+
+/// get a path to the file in given source location
+static std::string file(const source_locationt &l)
+{
+  if(l.get_file().empty())
+    return std::string();
+
+  const auto file = id2string(l.get_file());
+
+  if(source_locationt::is_built_in(file))
+    return std::string();
+
+  return concat_dir_file(id2string(l.get_working_directory()), file);
+}
+
+/// get the source files with non-top abstract states
+static std::set<std::string>
+get_source_files(const goto_modelt &goto_model, const ai_baset &ai)
+{
+  std::set<std::string> files;
+
+  for(const auto &f : goto_model.goto_functions.function_map)
+  {
+    forall_goto_program_instructions(i_it, f.second.body)
+    {
+      const auto abstract_state = ai.abstract_state_before(i_it);
+      if(abstract_state->is_top())
+        continue; // ignore
+
+      if(i_it->source_location.get_line().empty())
+        continue; // ignore
+
+      auto f = file(i_it->source_location);
+
+      if(!f.empty())
+        files.insert(f);
+    }
+  }
+
+  return files;
+}
+
+/// print a string with indent to match given sample line
+static void print_with_indent(
+  const std::string &src,
+  const std::string &indent_line,
+  std::ostream &out)
+{
+  const std::size_t p = indent_line.find_first_not_of(" \t");
+  const std::string indent =
+    p == std::string::npos ? std::string("") : std::string(indent_line, 0, p);
+  std::istringstream in(src);
+  std::string src_line;
+  while(std::getline(in, src_line))
+    out << indent << src_line << '\n';
+}
+
+/// output source code annotated with abstract states for given file
+void show_on_source(
+  const std::string &source_file,
+  const goto_modelt &goto_model,
+  const ai_baset &ai,
+  message_handlert &message_handler)
+{
+#ifdef _MSC_VER
+  std::ifstream in(widen(source_file));
+#else
+  std::ifstream in(source_file);
+#endif
+
+  if(!in)
+    return; // we fail silently
+
+  std::map<unsigned, ai_baset::locationt> line_map;
```

Comment:
I believe this should be `std::size_t` instead of `unsigned` - to me, the only reason to ever use `unsigned` are cases where memory consumption matters and we are damn sure it will be large enough.

Reply:
ok